[DDP] Use monitored barrier in DDP

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54742 test_c10d: use with_nccl_blocking_wait decorator
* #54741 test_c10d: Run tests with nccl_async_error_handling
* #54740 Provide a decorator to set/unset nccl blocking wait for tests
* #54558 [NCCL][Blocking Wait] Log set exceptions when checking for exceptions in
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* **#54220 [DDP] Use monitored barrier in DDP**
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* **#54220 [DDP] Use monitored barrier in DDP**
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* **#54220 [DDP] Use monitored barrier in DDP**
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* **#54220 [DDP] Use monitored barrier in DDP**
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier

Adds `dist.monitored_barrier()` call when after verifying model consistency across ranks to DDP constructor.

This is only enabled for debug modes > OFF, otherwise there is a slight perf hit during initialization. Will post benchmarks showing perf impact once they are done running.

Tested by ensuring the proper error propagation for the various debug modes when `dist.monitored_barrier()` is called.

Differential Revision: [D27106698](https://our.internmc.facebook.com/intern/diff/D27106698/)

rohan-varma committed:
[DDP] Use monitored barrier in DDP

Adds `dist.monitored_barrier()` call when after verifying model consistency across ranks to DDP constructor.

This is only enabled for debug modes > OFF, otherwise there is a slight perf hit during initialization. Will post benchmarks showing perf impact once they are done running.

Tested by ensuring the proper error propagation for the various debug modes when `dist.monitored_barrier()` is called.

Differential Revision: [D27106698](https://our.internmc.facebook.com/intern/diff/D27106698/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Use monitored barrier in DDP"


Adds `dist.monitored_barrier()` call when after verifying model consistency across ranks to DDP constructor.

This is only enabled for debug modes > OFF, otherwise there is a slight perf hit during initialization. Will post benchmarks showing perf impact once they are done running.

Tested by ensuring the proper error propagation for the various debug modes when `dist.monitored_barrier()` is called.

Differential Revision: [D27106698](https://our.internmc.facebook.com/intern/diff/D27106698/)

[ghstack-poisoned]

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -453,6 +459,9 @@ def __init__(self, module, device_ids=None,
 
+        # CPU-only process group used for debug synchronization in DDP when
+        # debug mode is enabled.
+        self._cpu_pg = self._create_gloo_pg()
```
could we move this line to after line 476?

rohan-varma replied:
This would be ideal, but it's challenging since some non-zero ranks may crash/hang at `dist._verify_model_across_ranks(self.process_group, parameters)` because of model mismatch. To successfully create the pg we need all ranks to call into it and having it after 476 could cause issues if some ranks crash early.

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -464,6 +473,11 @@ def __init__(self, module, device_ids=None,
+        if _GLOO_AVAILABLE and dist._get_debug_mode() != dist._DistributedDebugLevel.OFF:
+            # Run monitored barrier in debug mode. This will ensure all ranks
+            # have same model shapes before continuing with init/training.
+            dist.monitored_barrier(group=self._cpu_pg, timeout=30)
```
nit: Could we add something like LOG("calling  monitored_barrier after _verify_model_across_ranks....") before dist.monitored_barrier() call?

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -1313,6 +1329,22 @@ def _check_comm_hook(self, hook):
+        Creates a gloo process group with same ranks as self.process_group to be
+        used for debugability such as monitored_barrier().
+        """
+        if _GLOO_AVAILABLE and dist._get_debug_mode() != dist._DistributedDebugLevel.OFF:
```
feel like this condition can be removed so that _create_gloo_pg() can be reused for other cases

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -464,6 +473,11 @@ def __init__(self, module, device_ids=None,
+        if _GLOO_AVAILABLE and dist._get_debug_mode() != dist._DistributedDebugLevel.OFF:
+            # Run monitored barrier in debug mode. This will ensure all ranks
+            # have same model shapes before continuing with init/training.
+            dist.monitored_barrier(group=self._cpu_pg, timeout=30)
```
Are you planning to call dist.monitored_barrier after every collective call in DDP debug mode in following up diffs as well?

rohan-varma replied:
Yes, but the current plan for enabling that is only in DETAIL debug mode, whereas this one will run in INFO or DETAIL. I think it's worth having monitored barrier here for INFO as we only do it in DDP init, doing it for every collective comm. call in INFO mode as well is quite expensive IMO and should be saved for the most verbose debug option.

pritamdamania87 reviewed (commented):
Are we duplicating the monitored barrier here? I thought the ProcessGroupWrapper would by default call the monitored barrier before each collective. Isn't that sufficient instead of manually adding monitored barrier calls in debug mode?

I feel its more generic as well for the ProcessGroupWrapper to invoke monitored barrier before each collective since you automatically check all collective calls.

rohan-varma commented:
> Are we duplicating the monitored barrier here? I thought the ProcessGroupWrapper would by default call the monitored barrier before each collective. Isn't that sufficient instead of manually adding monitored barrier calls in debug mode?
> 
> I feel its more generic as well for the ProcessGroupWrapper to invoke monitored barrier before each collective since you automatically check all collective calls.

Agreed it is more generic and belongs in ProcessGroupWrapper. My plan for that was to only enable it in DETAIL debug mode as monitored barrier is quite expensive and blocking, calling it in every collective comm call will result in a lot of expected slowdown. In DDP here though I feel like having it in info mode as well will be useful as same model check has been a often-requested feature, and doing it in info mode at init time will not be too much of a perf hit. Thoughts?

rohan-varma committed:
Update on "[DDP] Use monitored barrier in DDP"


Adds `dist.monitored_barrier()` call when after verifying model consistency across ranks to DDP constructor.

This is only enabled for debug modes > OFF, otherwise there is a slight perf hit during initialization. Will post benchmarks showing perf impact once they are done running.

Tested by ensuring the proper error propagation for the various debug modes when `dist.monitored_barrier()` is called.

Differential Revision: [D27106698](https://our.internmc.facebook.com/intern/diff/D27106698/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Use monitored barrier in DDP"


Adds `dist.monitored_barrier()` call when after verifying model consistency across ranks to DDP constructor.

This is only enabled for debug modes > OFF, otherwise there is a slight perf hit during initialization. Will post benchmarks showing perf impact once they are done running.

Tested by ensuring the proper error propagation for the various debug modes when `dist.monitored_barrier()` is called.

Differential Revision: [D27106698](https://our.internmc.facebook.com/intern/diff/D27106698/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Use monitored barrier in DDP"


Adds `dist.monitored_barrier()` call when after verifying model consistency across ranks to DDP constructor.

This is only enabled for debug modes > OFF, otherwise there is a slight perf hit during initialization. Will post benchmarks showing perf impact once they are done running.

Tested by ensuring the proper error propagation for the various debug modes when `dist.monitored_barrier()` is called.

Differential Revision: [D27106698](https://our.internmc.facebook.com/intern/diff/D27106698/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Use monitored barrier in DDP"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* **#54220 [DDP] Use monitored barrier in DDP**
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier

Adds `dist.monitored_barrier()` call when after verifying model consistency across ranks to DDP constructor.

This is only enabled for debug modes > OFF, otherwise there is a slight perf hit during initialization. Will post benchmarks showing perf impact once they are done running.

Tested by ensuring the proper error propagation for the various debug modes when `dist.monitored_barrier()` is called.

Differential Revision: [D27106698](https://our.internmc.facebook.com/intern/diff/D27106698/)

[ghstack-poisoned]

pritamdamania87 commented:
> Agreed it is more generic and belongs in ProcessGroupWrapper. My plan for that was to only enable it in DETAIL debug mode as monitored barrier is quite expensive and blocking, calling it in every collective comm call will result in a lot of expected slowdown. In DDP here though I feel like having it in info mode as well will be useful as same model check has been a often-requested feature, and doing it in info mode at init time will not be too much of a perf hit. Thoughts?

I agree it is expensive and it should be enabled only for the highest debug mode. My feeling though is that we should consolidate the monitored barrier into a single debug mode via the ProcessGroupWrapper for consistency. Otherwise it is confusing to reason about when the monitored barrier is running. The way I see the debugging process work is set debug mode to info first to see if we can debug, if not continually increase the debug mode to see if additional information helps.

DDP does a bunch of collective calls, so the monitored barrier in this PR would probably only catch some of the issues. For example, after the monitored barrier we do `self._sync_params_and_buffers(authoritative_rank=0)` which might also run into some hangs/desync since it does a bunch of collective calls. I feel its cleaner just to ask users to use `DETAIL` level to debug any hangs.

rohan-varma commented:
@pritamdamania87 Sounds good, I'll abandon this PR for now and we can explore the proposal you suggested above.

Although, per your point about DDP doing a bunch of collective calls - this was the main motivation to have explicit calls to monitored_barrier() - the thinking is that with the wrapper pg, we will be able to identify the C++ collective that failed, but calls to monitored_barrier() in python seem more powerful, as we can notify the user what was going on in the program at a higher level. 

For example if we failed during model consistency check, pg wrapper, since it's implemented generically will just report an issue about failed call to broadcast(), but the underlying issue is that model is not consistent. If we add the explicit check here in python it seems that we can get to the true bug quicker. Thoughts on if theres a good way to do this?

pritamdamania87 commented:
> For example if we failed during model consistency check, pg wrapper, since it's implemented generically will just report an issue about failed call to broadcast(), but the underlying issue is that model is not consistent. If we add the explicit check here in python it seems that we can get to the true bug quicker. Thoughts on if theres a good way to do this?

Trying to understand the model consistency check here a bit more. Are we trying to validate that all ranks called into `_verify_model_across_ranks`? If all ranks called the DDP constructor, shouldn't they all enter that method unless some previous collective got stuck for some reason?

The thing is that even with this check here, when you actually enable the DETAIL debug mode, it will actually error out on the broadcast via the ProcessGroupWrapper and this check would never fire.

rohan-varma commented:
> Trying to understand the model consistency check here a bit more. Are we trying to validate that all ranks called into `_verify_model_across_ranks`? If all ranks called the DDP constructor, shouldn't they all enter that method unless some previous collective got stuck for some reason?

I discussed this with @zhaojuanmao offline, the gist was that we're not only trying to validate all ranks called `_verify_model_across_ranks`, but successfully passed it. Currently, if there is a model mismatch across ranks:
1. Rank 0 passes _verify_model_across_ranks
2. Rank 1 reports an exception and crashes
3. With NCCL backend, rank 0 eventually calls into some other DDP collective (hasn't crashed yet), and hangs in that collective.
4. Rank 0 is eventually taken down by nccl_async_error_handling or blocking wait, or just hangs

With this PR, this situation is improved because we call monitored_barrier() and rank 0 fails with a message indicating rank 1 did not cross the barrier, so the user can root cause the issue by checking rank 1's logs.

On the other hand, if we go with the process group wrapper approach, then rank 0 should also fail in the next collective because monitored_barrier() is called. However I believe that with this PR there is an opportunity to fail quicker in this scenario which could be beneficial.

> The thing is that even with this check here, when you actually enable the DETAIL debug mode, it will actually error out on the broadcast via the ProcessGroupWrapper and this check would never fire.

I guess it wouldn't error out in the broadcast in `_verify_model_across_ranks`, it would error out in some later collective call in DDP, so I think the check would fire right? 

Though overall, I do agree that it isn't best to insert these sort of checks after every DDP collective, we should come up with a more general solution. Open to either approach

rohan-varma committed:
Update on "[DDP] Use monitored barrier in DDP"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* **#54220 [DDP] Use monitored barrier in DDP**
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier

Adds `dist.monitored_barrier()` call when after verifying model consistency across ranks to DDP constructor.

This is only enabled for debug modes > OFF, otherwise there is a slight perf hit during initialization. Will post benchmarks showing perf impact once they are done running.

Tested by ensuring the proper error propagation for the various debug modes when `dist.monitored_barrier()` is called.

Differential Revision: [D27106698](https://our.internmc.facebook.com/intern/diff/D27106698/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Use monitored barrier in DDP"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* **#54220 [DDP] Use monitored barrier in DDP**
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* **#54220 [DDP] Use monitored barrier in DDP**
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier

Adds `dist.monitored_barrier()` call when after verifying model consistency across ranks to DDP constructor.

This is only enabled for debug modes > OFF, otherwise there is a slight perf hit during initialization. Will post benchmarks showing perf impact once they are done running.

Tested by ensuring the proper error propagation for the various debug modes when `dist.monitored_barrier()` is called.

Differential Revision: [D27106698](https://our.internmc.facebook.com/intern/diff/D27106698/)

[ghstack-poisoned]

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -1352,6 +1381,24 @@ def _check_comm_hook(self, hook):
+        """
+        if _GLOO_AVAILABLE:
+            ranks = list(range(dist.get_world_size(self.process_group)))
+            cpu_pg = dist.new_group(
```
actually I just realized that it is not good to call new_group() in DDP code base directly right now. as new_group() assumes users have created default process group using init_process_group(). but users does not have to call init_process_group() to use DDP, they could create a processGroup by directly using processGroupNCCL()...

so in case users are using processGroupNCCL() to create processGroup and pass it to DDP, calling new_group() will throw errors here.

zhaojuanmao reviewed (commented):
after second thought, i think I agree with you that calling monitored barrier possibly can make debugging faster. 

But I think we can hold this diff to be landed. After calling monitored_barrier before collective calls in processGroup wrapper, you can compare with it and see whether it really makes debugging faster.  e.g, possibly next collective call in DDP will be broadcast in sync_params, looks like it will not take too long to report the failure from rank 0. 

Also had a comment about using 'new_group' in DDP

rohan-varma commented:
Sounds good, we can shelve this change for now. 

> so in case users are using processGroupNCCL() to create processGroup and pass it to DDP, calling new_group() will throw errors here.

Do we support direct calls to creating `c10d.ProcessGroupNCCL` and passing that into DDP? I know we test that, but we don't documented it and I was under the impression that our process group creation apis are only `init_process_group` and `new_group`.

zhaojuanmao commented:
> Do we support direct calls to creating `c10d.ProcessGroupNCCL` and passing that into DDP? I know we test that, but we don't documented it and I was under the impression that our process group creation apis are only `init_process_group` and `new_group`.

yes, quite some unit tests in test_c10d directly call `c10d.ProcessGroupNCCL` and passing that into DDP
